stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -472,6 +473,13 @@ static bool sameObjCTypeManglings(Demangle::NodePointer node1,
 }
 #endif
 
+/// Optimization for the case where we need to compare a StringRef and a null terminated C string
+/// Not converting s2 to a StringRef avoids the need to call both strlen and memcmp
+static bool stringRefEqualsCString(StringRef s1, const char *s2) {
+  size_t length = s1.size();
+  return strncmp(s1.data(), s2, length) == 0 && s2[length] == '\0';
```

Comment:
Is there a possibility that `s1` could have an embedded `NUL` character? If so, the `strncmp` could find a match if `s2` matches the string up to the embedded `NUL`, but then `s2[length]` would be an out of bounds access.

Reply:
The code path is Byzantine, but this seems like a very remote possibility based on my current understanding.

Reply:
I worry that we could emit a symbolic reference into a context descriptor, at which point it's likely there _would_ be an embedded `NUL`. I haven't checked whether we do this today, but even if we don't do it now, we might do so in future.

Reply:
I believe the vast majority of these checks will be for the non-matching case, which means it should be nearly as fast to do:

```
  return strncmp(s1.data(), s2, length) == 0 && strlen(s2) == length;
```

We still have the double iteration in the case where they're actually equal, but that shouldn't noticeably impact overall performance.

Reply:
I think the matching case is equally important -- we see this overhead when dynamic casting like:

```
public protocol ErrorLoggerProtocol {
	func reportCrash()
}

public final class ErrorLogger: ErrorLoggerProtocol {
	public init() { }
	public func reportCrash() { }
}

func getLogger() -> Any {
	return ErrorLogger() as Any
}

var loggerProtocol: ErrorLoggerProtocol? = getLogger() as? ErrorLoggerProtocol
```

Contrived example -- the real code has a legitimate reason to want to do this. Oddly, if ErrorLoggerProtocol is annotated `@objc`, we see _better_ performance, which feels weird.

I don't have an amazing idea for solving this if the StringRef may have embedded `NUL` -- strncmp and strlen have optimized ARM implementations in libc, but I'm not aware of an off-the-shelf string comparator for this case.

Reply:
I've adopted the suggestion from @mikeash . Let me know if there's a useful way to benchmark the runtime, I'm happy to give it a whirl.